### DISCUSSION TITLE
Advanced facet filtering

### DISCUSSION THREAD

**Discussion by bobylito** [algolia/instantsearch#5736] [Ideas]
How to be able to handle 'or' filtering like `a=foo or b=bar`.

We could could let the user to set facetFilters like for the numeric refinement and the tags. The only problem, is then we avoid conflicts by not letting the user use methods such as add or remove until they clear the filters.

Opinions? @algolia/javascript

**Comment by olance** [comment]
Hope I'm not off-topic, but couldn't we accept an array of facets and consider these facets should be OR'd?

**Comment by vvo** [comment]
``` js
addFacetRefinement({
  attributeName: 'categories',
  value: 'tv'
  type: 'or'
});
```

this would a bit map what we propose in instantsearch.js

If I am also off-topic let me know :)

**Comment by olance** [comment]
hmm but what would it be OR'd with? The previous facet values? The next that will be added? Both?

**Comment by vvo** [comment]
Oh yes ahah tricky, type (or, and) of refinement is bound to context (other refinements in same facet).

So basically you would have to provide them all at once or default to and

**Comment by harrisminer** [comment]
Any updates on this issue?

**Comment by harrisminer** [comment]
bump

**Comment by olance** [comment]
Maybe the solution is to introduce a new abstraction to build such queries, much like what is done in ruby gems for SQL clients for instance.

A predicate object represents a boolean expression and can be combined with another predicate using or/and methods that return a new compound predicate for the new boolean expression.
Each predicate has a `toString()` method that outputs the actual filter string, a compound predicate will just join the inner predicates with the boolean operator it was created with.

``` javascript
// Returns a predicate object
let filter1 = algolia.filter('a = foo');
// Builds a compound predicate from the previous filter predicate
let filter2 = filter1.or('b = bar'); 
// Another compound from the previous compound predicate
let filter3 = filter2.and('c = baz'); 

filter1.toString(); // 'a = foo'
filter2.toString(); // 'a=foo OR b=bar'
filter3.toString(); // '(a=foo OR b=bar) AND c=baz'
```

Of course, one would rather write something like this:

``` javascript
search.setFilter(algolia.filter('a = foo').or('b = bar').and('c = baz'))
```

What do you think?

**Comment by olance** [comment]
In my idea, this would set the unified filter parameter and not mess with the facet/tagFilters.

**Comment by bobylito** [comment]
I like the concept of creating / managing with objects. It should make it possible to do complex refinements. I'm wondering how one can remove a predicate / filter?

> In my idea, this would set the unified filter parameter and not mess with the facet/tagFilters.

Does that mean you need to specify `_tags` when filtering?

That's a good start, let's iterate! (cc @pixelastic @vvo)

**Comment by olance** [comment]
@bobylito I'm not sure I get your question.
`_tags` only has to be used in your records if you want to use tag filters. But the "unified filter parameter" covers tags, facets and numeric refinements... so again, I'm not sure I get the question :)

As for removing predicates, it has to be thought carefully.
I think the right way to deal with this is to see the entire boolean expression as a tree. Taking my previous example:

```
         ______ AND ______
        |                 |
     _ OR _             c=baz
    |      |    
a=foo     b=bar
```

You might want to remove the "basic" predicates like `a=foo` or the compound ones like `a=foo OR b=bar`. Whenever you remove something, the task is to keep the tree balanced: removing a basic predicate means replacing its parent compound predicate by the other basic predicate that was left on the expression.  So removing `a=foo` makes the `OR` unbalanced, we replace it with `b=bar` and the full expression becomes `b=bar AND c=baz`.
It's actually the same thing when removing a full compound predicate. If you remove `a=foo OR b=bar` here, then `AND` becomes unbalanced and you have to remove it, leaving you with only `c=baz`.

So to sum this up, whenever we remove a predicate, the action is to replace its parent predicate by the other predicate of the expression. If it has no parent, just delete the predicate.

The question that remains to me is: **how do we address/identify those predicates?**

I think that covers everything?

**Comment by tensor-Overflow** [comment]
## ✅ Security Configuration Complete

### Final Security Updates Completed:
- **All Clerk/Stripe keys rotated**: New secret keys generated and deployed to Vercel/Render
- **CLERK_JWT_KEY configured**: Custom RSA private key generated and JWT template created in Clerk for enhanced session verification
- **Environment files secured**: Proper .gitignore patterns added to prevent future exposure

### Outstanding Items:
- **Clerk public key**: Cannot be regenerated (assessed as low security risk since it's designed to be public)
- **Environment separation**: All three environments (nt8-dev, mx-portal, mx-license) now properly isolated with secure configurations

### Security Status: 🔒 **HARDENED**
The licensing API now has enterprise-grade JWT verification with custom signing keys. All environments are properly secured and isolated.

---
**Next Phase**: Ready for integration development and production deployment with secure foundation in place.
